crates/deltalake-core/src/storage/mod.rs

```diff
@@ -22,11 +28,147 @@ lazy_static! {
     static ref DELTA_LOG_PATH: Path = Path::from("_delta_log");
 }
 
+/// Sharable reference to [`ObjectStore`]
+pub type ObjectStoreRef = Arc<DynObjectStore>;
+
+/// Factory trait for creating [ObjectStoreRef] instances at runtime
+pub trait ObjectStoreFactory: Send + Sync {
+    #[allow(missing_docs)]
+    fn parse_url_opts(
+        &self,
+        url: &Url,
+        options: &StorageOptions,
+    ) -> DeltaResult<(ObjectStoreRef, Path)>;
+}
+
+#[derive(Clone, Debug, Default)]
+struct DefaultObjectStoreFactory {}
+
+impl ObjectStoreFactory for DefaultObjectStoreFactory {
+    fn parse_url_opts(
+        &self,
+        url: &Url,
+        _options: &StorageOptions,
+    ) -> DeltaResult<(ObjectStoreRef, Path)> {
+        match url.scheme() {
+            "memory" => {
+                let path = Path::from_url_path(url.path())?;
+                let store = Arc::new(InMemory::new()) as ObjectStoreRef;
+                Ok((url_prefix_handler(store, path.clone())?, path))
+            }
+            "file" => {
+                let store = Arc::new(LocalFileSystem::new_with_prefix(
+                    url.to_file_path().unwrap(),
+                )?) as ObjectStoreRef;
+                Ok((store, Path::from("/")))
+            }
+            _ => Err(DeltaTableError::InvalidTableLocation(url.clone().into())),
+        }
+    }
+}
+
+/// TODO
+pub type FactoryRegistry = Arc<DashMap<Url, Arc<dyn ObjectStoreFactory>>>;
+
+/// TODO
+pub fn factories() -> FactoryRegistry {
```

Comment:
do we need to create global state for this? Ie. could we just place this on `DeltaTable`? 

My main concer is the integration with datafusion and other query engines, where such state is usually handled on a session. I also plan to to leverage the datafusion interns etc for our SCAN et. al. operations. In think in these cases having a global registry might limit hopw we can integrate.